diff --git a/third_party/vml/Lib/shared/array.vml b/third_party/vml/Lib/shared/array.vml
index d006a377..e5e1d717 100644
--- a/third_party/vml/Lib/shared/array.vml
+++ b/third_party/vml/Lib/shared/array.vml
@@ -1,7 +1,4 @@
 
-.linked "bitlib.vml"
-.linked "builtins.vml"
-.linked "util.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/array64.vml b/third_party/vml/Lib/shared/array64.vml
index d07df232..5dc60f06 100644
--- a/third_party/vml/Lib/shared/array64.vml
+++ b/third_party/vml/Lib/shared/array64.vml
@@ -1,6 +1,4 @@
 
-.linked "bitops64.vml"
-.linked "statistics64.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/base64.vml b/third_party/vml/Lib/shared/base64.vml
index b7b3ac44..28b02aa6 100644
--- a/third_party/vml/Lib/shared/base64.vml
+++ b/third_party/vml/Lib/shared/base64.vml
@@ -1,5 +1,4 @@
 
-.linked "string.vml"
 .data
 b64_table: .string "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
 
diff --git a/third_party/vml/Lib/shared/basiclib.vml b/third_party/vml/Lib/shared/basiclib.vml
index ecf39e53..7b305ff3 100644
--- a/third_party/vml/Lib/shared/basiclib.vml
+++ b/third_party/vml/Lib/shared/basiclib.vml
@@ -1,6 +1,4 @@
 
-.linked "math.vml"
-.linked "fixed.vml"
 .data
 _buf1: .word[256] 0
 L_66490000: .string "0123456789ABCDEF"
diff --git a/third_party/vml/Lib/shared/bitlib.vml b/third_party/vml/Lib/shared/bitlib.vml
index 8f191b83..5fa74d61 100644
--- a/third_party/vml/Lib/shared/bitlib.vml
+++ b/third_party/vml/Lib/shared/bitlib.vml
@@ -1,5 +1,4 @@
 
-.linked "builtins.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/browser_gfx.vml b/third_party/vml/Lib/shared/browser_gfx.vml
index 17b9e966..fded1a2b 100644
--- a/third_party/vml/Lib/shared/browser_gfx.vml
+++ b/third_party/vml/Lib/shared/browser_gfx.vml
@@ -1,5 +1,4 @@
 
-.linked "bitlib.vml"
 .data
 __cmd_offset: .word 0
 
diff --git a/third_party/vml/Lib/shared/builtins_basic.vml b/third_party/vml/Lib/shared/builtins_basic.vml
index 5f4cf115..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_basic.vml
+++ b/third_party/vml/Lib/shared/builtins_basic.vml
@@ -1,18 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "crt.vml"
-.linked "basiclib.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_c.vml b/third_party/vml/Lib/shared/builtins_c.vml
index 545306e0..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_c.vml
+++ b/third_party/vml/Lib/shared/builtins_c.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "printf.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_cpp.vml b/third_party/vml/Lib/shared/builtins_cpp.vml
index 545306e0..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_cpp.vml
+++ b/third_party/vml/Lib/shared/builtins_cpp.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "printf.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_csharp.vml b/third_party/vml/Lib/shared/builtins_csharp.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_csharp.vml
+++ b/third_party/vml/Lib/shared/builtins_csharp.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_d.vml b/third_party/vml/Lib/shared/builtins_d.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_d.vml
+++ b/third_party/vml/Lib/shared/builtins_d.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_dart.vml b/third_party/vml/Lib/shared/builtins_dart.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_dart.vml
+++ b/third_party/vml/Lib/shared/builtins_dart.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_forth.vml b/third_party/vml/Lib/shared/builtins_forth.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_forth.vml
+++ b/third_party/vml/Lib/shared/builtins_forth.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_fortran.vml b/third_party/vml/Lib/shared/builtins_fortran.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_fortran.vml
+++ b/third_party/vml/Lib/shared/builtins_fortran.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_go.vml b/third_party/vml/Lib/shared/builtins_go.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_go.vml
+++ b/third_party/vml/Lib/shared/builtins_go.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_java.vml b/third_party/vml/Lib/shared/builtins_java.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_java.vml
+++ b/third_party/vml/Lib/shared/builtins_java.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_javascript.vml b/third_party/vml/Lib/shared/builtins_javascript.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_javascript.vml
+++ b/third_party/vml/Lib/shared/builtins_javascript.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_kotlin.vml b/third_party/vml/Lib/shared/builtins_kotlin.vml
index 2e3ff2ab..43f3f716 100644
--- a/third_party/vml/Lib/shared/builtins_kotlin.vml
+++ b/third_party/vml/Lib/shared/builtins_kotlin.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_ladder.vml b/third_party/vml/Lib/shared/builtins_ladder.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_ladder.vml
+++ b/third_party/vml/Lib/shared/builtins_ladder.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_lua.vml b/third_party/vml/Lib/shared/builtins_lua.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_lua.vml
+++ b/third_party/vml/Lib/shared/builtins_lua.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_objc.vml b/third_party/vml/Lib/shared/builtins_objc.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_objc.vml
+++ b/third_party/vml/Lib/shared/builtins_objc.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_pascal.vml b/third_party/vml/Lib/shared/builtins_pascal.vml
index 0b193d3d..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_pascal.vml
+++ b/third_party/vml/Lib/shared/builtins_pascal.vml
@@ -1,17 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "crt.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_python.vml b/third_party/vml/Lib/shared/builtins_python.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_python.vml
+++ b/third_party/vml/Lib/shared/builtins_python.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_r.vml b/third_party/vml/Lib/shared/builtins_r.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_r.vml
+++ b/third_party/vml/Lib/shared/builtins_r.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_ruby.vml b/third_party/vml/Lib/shared/builtins_ruby.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_ruby.vml
+++ b/third_party/vml/Lib/shared/builtins_ruby.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_rust.vml b/third_party/vml/Lib/shared/builtins_rust.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_rust.vml
+++ b/third_party/vml/Lib/shared/builtins_rust.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_scheme.vml b/third_party/vml/Lib/shared/builtins_scheme.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_scheme.vml
+++ b/third_party/vml/Lib/shared/builtins_scheme.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/builtins_swift.vml b/third_party/vml/Lib/shared/builtins_swift.vml
index 1a428429..f57014ef 100644
--- a/third_party/vml/Lib/shared/builtins_swift.vml
+++ b/third_party/vml/Lib/shared/builtins_swift.vml
@@ -1,16 +1,4 @@
 
-.linked "file.vml"
-.linked "os.vml"
-.linked "convert.vml"
-.linked "string.vml"
-.linked "math.vml"
-.linked "printf.vml"
-.linked "io.vml"
-.linked "ctype.vml"
-.linked "bitops.vml"
-.linked "util.vml"
-.linked "float.vml"
-.linked "memory.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/button.vml b/third_party/vml/Lib/shared/button.vml
index d15d3335..771b5ae8 100644
--- a/third_party/vml/Lib/shared/button.vml
+++ b/third_party/vml/Lib/shared/button.vml
@@ -1,5 +1,4 @@
 
-.linked "swtimer.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/color.vml b/third_party/vml/Lib/shared/color.vml
index 5d55a55c..5d284177 100644
--- a/third_party/vml/Lib/shared/color.vml
+++ b/third_party/vml/Lib/shared/color.vml
@@ -767,6 +767,293 @@ L_27230066:
         pop R12
         pop R15
         ret
+hsv_to_rgb:
+; -------------------------------------------
+; source   : int hsv_to_rgb(int h,int s,int v)
+; function : hsv_to_rgb
+; param   : int h
+; param   : int s
+; param   : int v
+; return   : int
+; --------------------------------------------
+hsv_to_rgb:
+        push R15
+        push R12
+        move R12 R13
+        sub R13 #32
+        move R0 [R12+16]
+        push R0
+        move R0 #0
+        pop R1
+        cmp R1 R0
+        je _L27230069
+        move R0 #0
+        jmp _L27230070
+_L27230069:
+        move R0 #1
+_L27230070:
+        cmp R0 #0
+        jz sta_27230067
+        move R0 [R12+20]
+        push R0
+        move R0 [R12+20]
+        push R0
+        move R0 [R12+20]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230067:
+sta_27230068:
+        move R0 [R12+12]
+        push R0
+        move R0 #360
+        pop R1
+        mod R1 R0
+        move R0 R1
+        move [R12+12] R0
+        move R0 [R12+12]
+        push R0
+        move R0 #0
+        pop R1
+        cmp R1 R0
+        jl _L27230074
+        move R0 #0
+        jmp _L27230075
+_L27230074:
+        move R0 #1
+_L27230075:
+        cmp R0 #0
+        jz sta_27230072
+        move R0 [R12+12]
+        push R0
+        move R0 #360
+        pop R1
+        add R1 R0
+        move R0 R1
+        move [R12+12] R0
+sta_27230072:
+sta_27230073:
+        move R0 [R12+12]
+        push R0
+        move R0 #60
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-8] R0
+        move R0 [R12+12]
+        push R0
+        move R0 [R12-8]
+        push R0
+        move R0 #60
+        pop R1
+        mul R1 R0
+        move R0 R1
+        pop R1
+        sub R1 R0
+        move R0 R1
+        push R0
+        move R0 #255
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #60
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-12] R0
+        move R0 [R12+20]
+        push R0
+        move R0 #255
+        push R0
+        move R0 [R12+16]
+        pop R1
+        sub R1 R0
+        move R0 R1
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #255
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-16] R0
+        move R0 [R12+20]
+        push R0
+        move R0 #255
+        push R0
+        move R0 [R12+16]
+        push R0
+        move R0 [R12-12]
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #255
+        pop R1
+        div R1 R0
+        move R0 R1
+        pop R1
+        sub R1 R0
+        move R0 R1
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #255
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-20] R0
+        move R0 [R12+20]
+        push R0
+        move R0 #255
+        push R0
+        move R0 [R12+16]
+        push R0
+        move R0 #255
+        push R0
+        move R0 [R12-12]
+        pop R1
+        sub R1 R0
+        move R0 R1
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #255
+        pop R1
+        div R1 R0
+        move R0 R1
+        pop R1
+        sub R1 R0
+        move R0 R1
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #255
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-24] R0
+        move R0 [R12-8]
+        move R1 R0
+        push R1
+        move R0 #0
+        pop R1
+        cmp R0 R1
+        je sta_27230078
+        push R1
+        move R0 #1
+        pop R1
+        cmp R0 R1
+        je sta_27230079
+        push R1
+        move R0 #2
+        pop R1
+        cmp R0 R1
+        je sta_27230080
+        push R1
+        move R0 #3
+        pop R1
+        cmp R0 R1
+        je sta_27230081
+        push R1
+        move R0 #4
+        pop R1
+        cmp R0 R1
+        je sta_27230082
+        jmp sta_27230077
+sta_27230078:
+        move R0 [R12-16]
+        push R0
+        move R0 [R12-24]
+        push R0
+        move R0 [R12+20]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230079:
+        move R0 [R12-16]
+        push R0
+        move R0 [R12+20]
+        push R0
+        move R0 [R12-20]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230080:
+        move R0 [R12-24]
+        push R0
+        move R0 [R12+20]
+        push R0
+        move R0 [R12-16]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230081:
+        move R0 [R12+20]
+        push R0
+        move R0 [R12-20]
+        push R0
+        move R0 [R12-16]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230082:
+        move R0 [R12+20]
+        push R0
+        move R0 [R12-16]
+        push R0
+        move R0 [R12-24]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230077:
+        move R0 [R12-20]
+        push R0
+        move R0 [R12-16]
+        push R0
+        move R0 [R12+20]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        call rgb
+        add R13 #12
+        jmp L_27230071
+sta_27230076:
+L_27230071:
+        move R13 R12
+        pop R12
+        pop R15
+        ret
 color_black:
 ; -------------------------------------------
 ; source   : int color_black()
@@ -779,8 +1066,8 @@ color_black:
         move R12 R13
         sub R13 #12
         move R0 #0
-        jmp L_27230067
-L_27230067:
+        jmp L_27230083
+L_27230083:
         move R13 R12
         pop R12
         pop R15
@@ -797,8 +1084,8 @@ color_white:
         move R12 R13
         sub R13 #12
         move R0 #16777215
-        jmp L_27230068
-L_27230068:
+        jmp L_27230084
+L_27230084:
         move R13 R12
         pop R12
         pop R15
@@ -815,8 +1102,8 @@ color_red:
         move R12 R13
         sub R13 #12
         move R0 #16711680
-        jmp L_27230069
-L_27230069:
+        jmp L_27230085
+L_27230085:
         move R13 R12
         pop R12
         pop R15
@@ -833,8 +1120,8 @@ color_green:
         move R12 R13
         sub R13 #12
         move R0 #65280
-        jmp L_27230070
-L_27230070:
+        jmp L_27230086
+L_27230086:
         move R13 R12
         pop R12
         pop R15
@@ -851,8 +1138,8 @@ color_blue:
         move R12 R13
         sub R13 #12
         move R0 #255
-        jmp L_27230071
-L_27230071:
+        jmp L_27230087
+L_27230087:
         move R13 R12
         pop R12
         pop R15
@@ -869,8 +1156,8 @@ color_yellow:
         move R12 R13
         sub R13 #12
         move R0 #16776960
-        jmp L_27230072
-L_27230072:
+        jmp L_27230088
+L_27230088:
         move R13 R12
         pop R12
         pop R15
@@ -887,8 +1174,8 @@ color_cyan:
         move R12 R13
         sub R13 #12
         move R0 #65535
-        jmp L_27230073
-L_27230073:
+        jmp L_27230089
+L_27230089:
         move R13 R12
         pop R12
         pop R15
@@ -906,8 +1193,8 @@ color_magenta:
         move R12 R13
         sub R13 #12
         move R0 #16711935
-        jmp L_27230074
-L_27230074:
+        jmp L_27230090
+L_27230090:
         move R13 R12
         pop R12
         pop R15
diff --git a/third_party/vml/Lib/shared/complex.vml b/third_party/vml/Lib/shared/complex.vml
index 2e637604..dab6419a 100644
--- a/third_party/vml/Lib/shared/complex.vml
+++ b/third_party/vml/Lib/shared/complex.vml
@@ -1,6 +1,4 @@
 
-.linked "fixed.vml"
-.linked "math.vml"
 .data
 flt_43840004: .word 786163455
 flt_43840007: .word 786163455
diff --git a/third_party/vml/Lib/shared/complex64.vml b/third_party/vml/Lib/shared/complex64.vml
index f04d00f8..c2a1cae1 100644
--- a/third_party/vml/Lib/shared/complex64.vml
+++ b/third_party/vml/Lib/shared/complex64.vml
@@ -1,5 +1,4 @@
 
-.linked "math.vml"
 .data
 dbl_92980002: .dword 0
 dbl_92980005: .dword 0
diff --git a/third_party/vml/Lib/shared/convert.vml b/third_party/vml/Lib/shared/convert.vml
index 0c93776d..846f183d 100644
--- a/third_party/vml/Lib/shared/convert.vml
+++ b/third_party/vml/Lib/shared/convert.vml
@@ -1,3 +1,4 @@
+
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/crosslang.vml b/third_party/vml/Lib/shared/crosslang.vml
index d5086217..e1bbd20b 100644
--- a/third_party/vml/Lib/shared/crosslang.vml
+++ b/third_party/vml/Lib/shared/crosslang.vml
@@ -1,6 +1,4 @@
 
-.linked "fixed.vml"
-.linked "math.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/crosslang64.vml b/third_party/vml/Lib/shared/crosslang64.vml
index b228891d..14ac2904 100644
--- a/third_party/vml/Lib/shared/crosslang64.vml
+++ b/third_party/vml/Lib/shared/crosslang64.vml
@@ -1,5 +1,4 @@
 
-.linked "statistics64.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/ctype.vml b/third_party/vml/Lib/shared/ctype.vml
index 4ffa32ce..0cbb0419 100644
--- a/third_party/vml/Lib/shared/ctype.vml
+++ b/third_party/vml/Lib/shared/ctype.vml
@@ -1,5 +1,4 @@
 
-.linked "parserexpf.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/device64.vml b/third_party/vml/Lib/shared/device64.vml
index 3674dfec..16a684c6 100644
--- a/third_party/vml/Lib/shared/device64.vml
+++ b/third_party/vml/Lib/shared/device64.vml
@@ -1,5 +1,4 @@
 
-.linked "device.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/encoding64.vml b/third_party/vml/Lib/shared/encoding64.vml
index 42b94bc7..2ab575b8 100644
--- a/third_party/vml/Lib/shared/encoding64.vml
+++ b/third_party/vml/Lib/shared/encoding64.vml
@@ -1,5 +1,4 @@
 
-.linked "encoding.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/fixed.vml b/third_party/vml/Lib/shared/fixed.vml
index c4737242..8008454a 100644
--- a/third_party/vml/Lib/shared/fixed.vml
+++ b/third_party/vml/Lib/shared/fixed.vml
@@ -1,5 +1,4 @@
 
-.linked "math.vml"
 .data
 flt_57480009: .word 1199570944
 flt_57480011: .word 1199570944
diff --git a/third_party/vml/Lib/shared/fixed64.vml b/third_party/vml/Lib/shared/fixed64.vml
index 0410cc8a..cb0a1c5e 100644
--- a/third_party/vml/Lib/shared/fixed64.vml
+++ b/third_party/vml/Lib/shared/fixed64.vml
@@ -2,6 +2,7 @@
 .data
 i64_49580025: .dword 1
 i64_49580027: .dword 1
+i64_49580029: .dword 1
 
 .text
 lfixed_mul64:
@@ -338,6 +339,36 @@ L_49580026:
         pop R12
         pop R15
         ret
+lfixed_from_double:
+; -------------------------------------------
+; source   : long lfixed_from_double(double f)
+; function : lfixed_from_double
+; param   : double f
+; return   : long
+; --------------------------------------------
+lfixed_from_double:
+        push R15
+        push R12
+        move R12 R13
+        sub R13 #12
+        moved R0 [R12+12]
+        dpush R0
+        movel R0 [i64_49580027]
+        pushl R0
+        move R0 #32
+        popl R1
+        shl R1 R0
+        movel R0 R1
+        l2d R0 R0
+        dpop R1
+        dmul R0 R1 R0
+        d2l R0 R0
+        jmp L_49580028
+L_49580028:
+        move R13 R12
+        pop R12
+        pop R15
+        ret
 lfixed_to_double:
 ; -------------------------------------------:
 ; -------------------------------------------
@@ -354,7 +385,7 @@ lfixed_to_double:
         movel R0 [R12+12]
         l2d R0 R0
         dpush R0
-        movel R0 [i64_49580027]
+        movel R0 [i64_49580029]
         pushl R0
         move R0 #32
         popl R1
@@ -363,8 +394,8 @@ lfixed_to_double:
         l2d R0 R0
         dpop R1
         ddiv R0 R1 R0
-        jmp L_49580028
-L_49580028:
+        jmp L_49580030
+L_49580030:
         move R13 R12
         pop R12
         pop R15
diff --git a/third_party/vml/Lib/shared/fsm.vml b/third_party/vml/Lib/shared/fsm.vml
index 5bde7e8c..9a0ddf1d 100644
--- a/third_party/vml/Lib/shared/fsm.vml
+++ b/third_party/vml/Lib/shared/fsm.vml
@@ -1,5 +1,4 @@
 
-.linked "scheduler.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/graph.vml b/third_party/vml/Lib/shared/graph.vml
index d6396146..f18bcc50 100644
--- a/third_party/vml/Lib/shared/graph.vml
+++ b/third_party/vml/Lib/shared/graph.vml
@@ -1,6 +1,4 @@
 
-.linked "array.vml"
-.linked "math.vml"
 .data
 _bgi_color: .word 15
 _bgi_bgcolor: .word 0
diff --git a/third_party/vml/Lib/shared/graphics.vml b/third_party/vml/Lib/shared/graphics.vml
index f76414e1..a4acb847 100644
--- a/third_party/vml/Lib/shared/graphics.vml
+++ b/third_party/vml/Lib/shared/graphics.vml
@@ -1,6 +1,4 @@
 
-.linked "array.vml"
-.linked "math.vml"
 .data
 _gfx_fb: .word 0
 _gfx_w: .word 0
@@ -5053,10 +5051,10 @@ sta_69550301:
         pop R12
         pop R15
         ret
-gfx_sector_fill:
+gfx_sector:
 ; -------------------------------------------
-; source   : void gfx_sector_fill(int cx,int cy,int r,int start_angle,int end_angle,int color)
-; function : gfx_sector_fill
+; source   : void gfx_sector(int cx,int cy,int r,int start_angle,int end_angle,int color)
+; function : gfx_sector
 ; param   : int cx
 ; param   : int cy
 ; param   : int r
@@ -5065,7 +5063,7 @@ gfx_sector_fill:
 ; param   : int color
 ; return   : void
 ; --------------------------------------------
-gfx_sector_fill:
+gfx_sector:
         push R15
         push R12
         move R12 R13
@@ -5108,18 +5106,97 @@ _L69550316:
         cmp R0 #0
         jz sta_69550314
         move R0 [R12-8]
+        move [R12-16] R0
+        move R0 [R12+12]
         push R0
-        move R0 #360
+        move R0 [R12+20]
+        push R0
+        move R0 #90
+        push R0
+        move R0 [R12-16]
         pop R1
-        mod R1 R0
+        sub R1 R0
         move R0 R1
-        move [R12-16] R0
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #90
+        pop R1
+        div R1 R0
+        move R0 R1
+        pop R1
+        add R1 R0
+        move R0 R1
+        move [R12-20] R0
+        move R0 [R12+16]
+        push R0
+        move R0 [R12+20]
+        push R0
         move R0 [R12-16]
+        pop R1
+        mul R1 R0
+        move R0 R1
         push R0
         move R0 #90
         pop R1
+        div R1 R0
+        move R0 R1
+        pop R1
+        add R1 R0
+        move R0 R1
+        move [R12-24] R0
+        move R0 [R12+32]
+        push R0
+        move R0 [R12-24]
+        push R0
+        move R0 [R12-20]
+        push R0
+        move R0 [R12+16]
+        push R0
+        move R0 [R12+12]
+        push R0
+        move R0 [R13+0]
+        move R1 [R13+4]
+        move R2 [R13+8]
+        move R3 [R13+12]
+        call gfx_line
+        add R13 #20
+sta_69550313:
+        move R0 [R12-8]
+        move R1 R0
+        add R0 R0 #1
+        move [R12-8] R0
+        move R0 R1
+        jmp sta_69550312
+sta_69550314:
+        move R13 R12
+        pop R12
+        pop R15
+        ret
+gfx_sector_fill:
+; -------------------------------------------
+; source   : void gfx_sector_fill(int cx,int cy,int r,int start_angle,int end_angle,int color)
+; function : gfx_sector_fill
+; param   : int cx
+; param   : int cy
+; param   : int r
+; param   : int start_angle
+; param   : int end_angle
+; param   : int color
+; return   : void
+; --------------------------------------------
+gfx_sector_fill:
+        push R15
+        push R12
+        move R12 R13
+        sub R13 #32
+        move R0 [R12+24]
+        push R0
+        move R0 [R12+28]
+        pop R1
         cmp R1 R0
-        jle _L69550319
+        jg _L69550319
         move R0 #0
         jmp _L69550320
 _L69550319:
@@ -5127,6 +5204,50 @@ _L69550319:
 _L69550320:
         cmp R0 #0
         jz sta_69550317
+        move R0 [R12+24]
+        move [R12-12] R0
+        move R0 [R12+28]
+        move [R12+24] R0
+        move R0 [R12-12]
+        move [R12+28] R0
+sta_69550317:
+sta_69550318:
+        move R0 [R12+24]
+        move [R12-8] R0
+sta_69550321:
+        move R0 [R12-8]
+        push R0
+        move R0 [R12+28]
+        pop R1
+        cmp R1 R0
+        jle _L69550324
+        move R0 #0
+        jmp _L69550325
+_L69550324:
+        move R0 #1
+_L69550325:
+        cmp R0 #0
+        jz sta_69550323
+        move R0 [R12-8]
+        push R0
+        move R0 #360
+        pop R1
+        mod R1 R0
+        move R0 R1
+        move [R12-16] R0
+        move R0 [R12-16]
+        push R0
+        move R0 #90
+        pop R1
+        cmp R1 R0
+        jle _L69550328
+        move R0 #0
+        jmp _L69550329
+_L69550328:
+        move R0 #1
+_L69550329:
+        cmp R0 #0
+        jz sta_69550326
         move R0 #90
         push R0
         move R0 [R12-16]
@@ -5156,21 +5277,21 @@ _L69550320:
         div R1 R0
         move R0 R1
         move [R12-24] R0
-        jmp sta_69550318
-sta_69550317:
+        jmp sta_69550327
+sta_69550326:
         move R0 [R12-16]
         push R0
         move R0 #180
         pop R1
         cmp R1 R0
-        jle _L69550323
+        jle _L69550332
         move R0 #0
-        jmp _L69550324
-_L69550323:
+        jmp _L69550333
+_L69550332:
         move R0 #1
-_L69550324:
+_L69550333:
         cmp R0 #0
-        jz sta_69550321
+        jz sta_69550330
         move R0 [R12-16]
         push R0
         move R0 #90
@@ -5212,21 +5333,21 @@ _L69550324:
         div R1 R0
         move R0 R1
         move [R12-24] R0
-        jmp sta_69550322
-sta_69550321:
+        jmp sta_69550331
+sta_69550330:
         move R0 [R12-16]
         push R0
         move R0 #270
         pop R1
         cmp R1 R0
-        jle _L69550327
+        jle _L69550336
         move R0 #0
-        jmp _L69550328
-_L69550327:
+        jmp _L69550337
+_L69550336:
         move R0 #1
-_L69550328:
+_L69550337:
         cmp R0 #0
-        jz sta_69550325
+        jz sta_69550334
         move R0 #270
         push R0
         move R0 [R12-16]
@@ -5275,8 +5396,8 @@ _L69550328:
         div R1 R0
         move R0 R1
         move [R12-24] R0
-        jmp sta_69550326
-sta_69550325:
+        jmp sta_69550335
+sta_69550334:
         move R0 [R12-16]
         push R0
         move R0 #270
@@ -5318,9 +5439,9 @@ sta_69550325:
         div R1 R0
         move R0 R1
         move [R12-24] R0
-sta_69550326:
-sta_69550322:
-sta_69550318:
+sta_69550335:
+sta_69550331:
+sta_69550327:
         move R0 [R12+32]
         push R0
         move R0 [R12+16]
@@ -5347,7 +5468,7 @@ sta_69550318:
         move R3 [R13+12]
         call gfx_line
         add R13 #20
-sta_69550313:
+sta_69550322:
         move R0 [R12-8]
         push R0
         move R0 #2
@@ -5355,8 +5476,8 @@ sta_69550313:
         add R1 R0
         move R0 R1
         move [R12-8] R0
-        jmp sta_69550312
-sta_69550314:
+        jmp sta_69550321
+sta_69550323:
         move R0 [R12+32]
         push R0
         move R0 [R12+28]
@@ -5438,7 +5559,7 @@ gfx_print_scale:
         move [R12-16] R0
         move R0 #0
         move [R12-20] R0
-sta_69550329:
+sta_69550338:
         move R2 #0
         move R0 [R12+20]
         push R0
@@ -5452,14 +5573,14 @@ sta_69550329:
         move R0 #0
         popb R1
         cmp R1 R0
-        jne _L69550331
+        jne _L69550340
         move R0 #0
-        jmp _L69550332
-_L69550331:
+        jmp _L69550341
+_L69550340:
         move R0 #1
-_L69550332:
+_L69550341:
         cmp R0 #0
-        jz sta_69550330
+        jz sta_69550339
         move R0 [R12+20]
         push R0
         move R0 [R12-20]
@@ -5475,14 +5596,14 @@ _L69550332:
         move R0 #10
         popb R1
         cmp R1 R0
-        je _L69550335
+        je _L69550344
         move R0 #0
-        jmp _L69550336
-_L69550335:
+        jmp _L69550345
+_L69550344:
         move R0 #1
-_L69550336:
+_L69550345:
         cmp R0 #0
-        jz sta_69550333
+        jz sta_69550342
         move R0 [R12-8]
         move [R12-12] R0
         move R0 [R12-16]
@@ -5502,49 +5623,49 @@ _L69550336:
         add R0 R0 #1
         move [R12-20] R0
         move R0 R1
-        jmp sta_69550329
-sta_69550333:
-sta_69550334:
+        jmp sta_69550338
+sta_69550342:
+sta_69550343:
         moveb R0 [R12-21]
         sextb R0 R0
         push R0
         move R0 #32
         pop R1
         cmp R1 R0
-        jl _L69550342
+        jl _L69550351
         move R0 #0
-        jmp _L69550343
-_L69550342:
+        jmp _L69550352
+_L69550351:
         move R0 #1
-_L69550343:
+_L69550352:
         test R0 R0
-        jnz _L69550339
+        jnz _L69550348
         moveb R0 [R12-21]
         sextb R0 R0
         push R0
         move R0 #126
         pop R1
         cmp R1 R0
-        jg _L69550344
+        jg _L69550353
         move R0 #0
-        jmp _L69550345
-_L69550344:
+        jmp _L69550354
+_L69550353:
         move R0 #1
-_L69550345:
+_L69550354:
         test R0 R0
-        jz _L69550340
-_L69550339:
+        jz _L69550349
+_L69550348:
         move R0 #1
-        jmp _L69550341
-_L69550340:
+        jmp _L69550350
+_L69550349:
         move R0 #0
-_L69550341:
+_L69550350:
         cmp R0 #0
-        jz sta_69550337
+        jz sta_69550346
         move R0 #32
         moveb [R12-21] R0
-sta_69550337:
-sta_69550338:
+sta_69550346:
+sta_69550347:
         moveb R0 [R12-21]
         sextb R0 R0
         push R0
@@ -5560,20 +5681,20 @@ sta_69550338:
         move [R12-28] R0
         move R0 #0
         move [R12-32] R0
-sta_69550346:
+sta_69550355:
         move R0 [R12-32]
         push R0
         move R0 #16
         pop R1
         cmp R1 R0
-        jl _L69550349
+        jl _L69550358
         move R0 #0
-        jmp _L69550350
-_L69550349:
+        jmp _L69550359
+_L69550358:
         move R0 #1
-_L69550350:
+_L69550359:
         cmp R0 #0
-        jz sta_69550348
+        jz sta_69550357
         move R0 font8x16
         push R0
         move R9 #0
@@ -5594,20 +5715,20 @@ _L69550350:
         moveb [R12-37] R0
         move R0 #0
         move [R12-36] R0
-sta_69550351:
+sta_69550360:
         move R0 [R12-36]
         push R0
         move R0 #8
         pop R1
         cmp R1 R0
-        jl _L69550354
+        jl _L69550363
         move R0 #0
-        jmp _L69550355
-_L69550354:
+        jmp _L69550364
+_L69550363:
         move R0 #1
-_L69550355:
+_L69550364:
         cmp R0 #0
-        jz sta_69550353
+        jz sta_69550362
         moveb R0 [R12-37]
         push R0
         move R0 #128
@@ -5620,39 +5741,39 @@ _L69550355:
         and R1 R0
         move R0 R1
         cmp R0 #0
-        jz sta_69550356
+        jz sta_69550365
         move R0 #0
         move [R12-44] R0
-sta_69550358:
+sta_69550367:
         move R0 [R12-44]
         push R0
         move R0 [R12+28]
         pop R1
         cmp R1 R0
-        jl _L69550361
+        jl _L69550370
         move R0 #0
-        jmp _L69550362
-_L69550361:
+        jmp _L69550371
+_L69550370:
         move R0 #1
-_L69550362:
+_L69550371:
         cmp R0 #0
-        jz sta_69550360
+        jz sta_69550369
         move R0 #0
         move [R12-48] R0
-sta_69550363:
+sta_69550372:
         move R0 [R12-48]
         push R0
         move R0 [R12+28]
         pop R1
         cmp R1 R0
-        jl _L69550366
+        jl _L69550375
         move R0 #0
-        jmp _L69550367
-_L69550366:
+        jmp _L69550376
+_L69550375:
         move R0 #1
-_L69550367:
+_L69550376:
         cmp R0 #0
-        jz sta_69550365
+        jz sta_69550374
         move R0 [R12+24]
         push R0
         move R0 [R12-16]
@@ -5694,40 +5815,40 @@ _L69550367:
         move R2 [R13+8]
         call _putpixel
         add R13 #12
-sta_69550364:
+sta_69550373:
         move R0 [R12-48]
         move R1 R0
         add R0 R0 #1
         move [R12-48] R0
         move R0 R1
-        jmp sta_69550363
-sta_69550365:
-sta_69550359:
+        jmp sta_69550372
+sta_69550374:
+sta_69550368:
         move R0 [R12-44]
         move R1 R0
         add R0 R0 #1
         move [R12-44] R0
         move R0 R1
-        jmp sta_69550358
-sta_69550360:
-sta_69550356:
-sta_69550357:
-sta_69550352:
+        jmp sta_69550367
+sta_69550369:
+sta_69550365:
+sta_69550366:
+sta_69550361:
         move R0 [R12-36]
         move R1 R0
         add R0 R0 #1
         move [R12-36] R0
         move R0 R1
-        jmp sta_69550351
-sta_69550353:
-sta_69550347:
+        jmp sta_69550360
+sta_69550362:
+sta_69550356:
         move R0 [R12-32]
         move R1 R0
         add R0 R0 #1
         move [R12-32] R0
         move R0 R1
-        jmp sta_69550346
-sta_69550348:
+        jmp sta_69550355
+sta_69550357:
         move R0 [R12-12]
         push R0
         move R0 #8
@@ -5745,8 +5866,8 @@ sta_69550348:
         add R0 R0 #1
         move [R12-20] R0
         move R0 R1
-        jmp sta_69550329
-sta_69550330:
+        jmp sta_69550338
+sta_69550339:
         move R13 R12
         pop R12
         pop R15
@@ -5786,38 +5907,38 @@ gfx_flood_fill:
         move R0 [R12+20]
         pop R1
         cmp R1 R0
-        je _L69550373
+        je _L69550382
         move R0 #0
-        jmp _L69550374
-_L69550373:
+        jmp _L69550383
+_L69550382:
         move R0 #1
-_L69550374:
+_L69550383:
         test R0 R0
-        jnz _L69550370
+        jnz _L69550379
         move R0 [R12-20]
         push R0
         move R0 [R12+24]
         pop R1
         cmp R1 R0
-        je _L69550375
+        je _L69550384
         move R0 #0
-        jmp _L69550376
-_L69550375:
+        jmp _L69550385
+_L69550384:
         move R0 #1
-_L69550376:
+_L69550385:
         test R0 R0
-        jz _L69550371
-_L69550370:
+        jz _L69550380
+_L69550379:
         move R0 #1
-        jmp _L69550372
-_L69550371:
+        jmp _L69550381
+_L69550380:
         move R0 #0
-_L69550372:
+_L69550381:
         cmp R0 #0
-        jz sta_69550368
-        jmp L_69550377
-sta_69550368:
-sta_69550369:
+        jz sta_69550377
+        jmp L_69550386
+sta_69550377:
+sta_69550378:
         move R0 [R12+12]
         move R7 R0
         move R0 _flood_sx
@@ -5851,41 +5972,41 @@ sta_69550369:
         add R0 R0 #1
         move [R12-8] R0
         move R0 R1
-sta_69550378:
+sta_69550387:
         move R2 #0
         move R0 [R12-8]
         push R0
         move R0 #0
         pop R1
         cmp R1 R0
-        jg _L69550382
+        jg _L69550391
         move R0 #0
-        jmp _L69550383
-_L69550382:
+        jmp _L69550392
+_L69550391:
         move R0 #1
-_L69550383:
+_L69550392:
         test R0 R0
-        jz _L69550380
+        jz _L69550389
         move R0 [R12-8]
         push R0
         move R0 #256
         pop R1
         cmp R1 R0
-        jl _L69550384
+        jl _L69550393
         move R0 #0
-        jmp _L69550385
-_L69550384:
+        jmp _L69550394
+_L69550393:
         move R0 #1
-_L69550385:
+_L69550394:
         test R0 R0
-        jz _L69550380
+        jz _L69550389
         move R0 #1
-        jmp _L69550381
-_L69550380:
+        jmp _L69550390
+_L69550389:
         move R0 #0
-_L69550381:
+_L69550390:
         cmp R0 #0
-        jz sta_69550379
+        jz sta_69550388
         move R0 [R12-8]
         move R1 R0
         sub R0 R0 #1
@@ -5922,80 +6043,80 @@ _L69550381:
         move R0 #0
         pop R1
         cmp R1 R0
-        jl _L69550397
+        jl _L69550406
         move R0 #0
-        jmp _L69550398
-_L69550397:
+        jmp _L69550407
+_L69550406:
         move R0 #1
-_L69550398:
+_L69550407:
         test R0 R0
-        jnz _L69550394
+        jnz _L69550403
         move R0 [R12-24]
         push R0
         move R0 [R12-12]
         pop R1
         cmp R1 R0
-        jge _L69550399
+        jge _L69550408
         move R0 #0
-        jmp _L69550400
-_L69550399:
+        jmp _L69550409
+_L69550408:
         move R0 #1
-_L69550400:
+_L69550409:
         test R0 R0
-        jz _L69550395
-_L69550394:
+        jz _L69550404
+_L69550403:
         move R0 #1
-        jmp _L69550396
-_L69550395:
+        jmp _L69550405
+_L69550404:
         move R0 #0
-_L69550396:
+_L69550405:
         test R0 R0
-        jnz _L69550391
+        jnz _L69550400
         move R0 [R12-28]
         push R0
         move R0 #0
         pop R1
         cmp R1 R0
-        jl _L69550401
+        jl _L69550410
         move R0 #0
-        jmp _L69550402
-_L69550401:
+        jmp _L69550411
+_L69550410:
         move R0 #1
-_L69550402:
+_L69550411:
         test R0 R0
-        jz _L69550392
-_L69550391:
+        jz _L69550401
+_L69550400:
         move R0 #1
-        jmp _L69550393
-_L69550392:
+        jmp _L69550402
+_L69550401:
         move R0 #0
-_L69550393:
+_L69550402:
         test R0 R0
-        jnz _L69550388
+        jnz _L69550397
         move R0 [R12-28]
         push R0
         move R0 [R12-16]
         pop R1
         cmp R1 R0
-        jge _L69550403
+        jge _L69550412
         move R0 #0
-        jmp _L69550404
-_L69550403:
+        jmp _L69550413
+_L69550412:
         move R0 #1
-_L69550404:
+_L69550413:
         test R0 R0
-        jz _L69550389
-_L69550388:
+        jz _L69550398
+_L69550397:
         move R0 #1
-        jmp _L69550390
-_L69550389:
+        jmp _L69550399
+_L69550398:
         move R0 #0
-_L69550390:
+_L69550399:
         cmp R0 #0
-        jz sta_69550386
-        jmp sta_69550378
-sta_69550386:
-sta_69550387:
+        jz sta_69550395
+        jmp sta_69550387
+sta_69550395:
+sta_69550396:
         move R0 [R12-28]
         push R0
         move R0 [R12-24]
@@ -6008,17 +6129,17 @@ sta_69550387:
         move R0 [R12-20]
         pop R1
         cmp R1 R0
-        jne _L69550407
+        jne _L69550416
         move R0 #0
-        jmp _L69550408
-_L69550407:
+        jmp _L69550417
+_L69550416:
         move R0 #1
-_L69550408:
+_L69550417:
         cmp R0 #0
-        jz sta_69550405
-        jmp sta_69550378
-sta_69550405:
-sta_69550406:
+        jz sta_69550414
+        jmp sta_69550387
+sta_69550414:
+sta_69550415:
         move R0 [R12+20]
         push R0
         move R0 [R12-28]
@@ -6040,14 +6161,14 @@ sta_69550406:
         move R0 #256
         pop R1
         cmp R1 R0
-        jl _L69550411
+        jl _L69550420
         move R0 #0
-        jmp _L69550412
-_L69550411:
+        jmp _L69550421
+_L69550420:
         move R0 #1
-_L69550412:
+_L69550421:
         cmp R0 #0
-        jz sta_69550409
+        jz sta_69550418
         move R0 [R12-24]
         push R0
         move R0 #1
@@ -6200,11 +6321,11 @@ _L69550412:
         add R0 R0 #1
         move [R12-8] R0
         move R0 R1
-sta_69550409:
-sta_69550410:
-        jmp sta_69550378
-sta_69550379:
-L_69550377:
+sta_69550418:
+sta_69550419:
+        jmp sta_69550387
+sta_69550388:
+L_69550386:
         move R13 R12
         pop R12
         pop R15
@@ -6223,8 +6344,8 @@ gfx_screenshot:
         asm "SYSCALL 200"
         asm "STORE R0, _vml_result_int"
         move R0 [_vml_result_int]
-        jmp L_69550413
-L_69550413:
+        jmp L_69550422
+L_69550422:
         move R13 R12
         pop R12
         pop R15
@@ -6248,8 +6369,8 @@ gfx_put_image:
         asm "SYSCALL 201"
         asm "STORE R0, _vml_result_int"
         move R0 [_vml_result_int]
-        jmp L_69550414
-L_69550414:
+        jmp L_69550423
+L_69550423:
         move R13 R12
         pop R12
         pop R15
@@ -6273,8 +6394,8 @@ gfx_get_image:
         asm "SYSCALL 202"
         asm "STORE R0, _vml_result_int"
         move R0 [_vml_result_int]
-        jmp L_69550415
-L_69550415:
+        jmp L_69550424
+L_69550424:
         move R13 R12
         pop R12
         pop R15
@@ -6298,8 +6419,8 @@ gfx_viewport:
         asm "SYSCALL 203"
         asm "STORE R0, _vml_result_int"
         move R0 [_vml_result_int]
-        jmp L_69550416
-L_69550416:
+        jmp L_69550425
+L_69550425:
         move R13 R12
         pop R12
         pop R15
diff --git a/third_party/vml/Lib/shared/io.vml b/third_party/vml/Lib/shared/io.vml
index 08f9377f..7f1990c6 100644
--- a/third_party/vml/Lib/shared/io.vml
+++ b/third_party/vml/Lib/shared/io.vml
@@ -1,12 +1,4 @@
 
-.linked "builtins.vml"
-.linked "wchar.vml"
-.linked "uchar.vml"
-.linked "ctype.vml"
-.linked "printf.vml"
-.linked "uscanf.vml"
-.linked "wscanf.vml"
-.linked "wstring.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/io64.vml b/third_party/vml/Lib/shared/io64.vml
index 9e21a1ea..edf877c1 100644
--- a/third_party/vml/Lib/shared/io64.vml
+++ b/third_party/vml/Lib/shared/io64.vml
@@ -1,7 +1,4 @@
 
-.linked "convert64.vml"
-.linked "io.vml"
-.linked "printf.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/math.vml b/third_party/vml/Lib/shared/math.vml
index d8408450..f3e7dbb3 100644
--- a/third_party/vml/Lib/shared/math.vml
+++ b/third_party/vml/Lib/shared/math.vml
@@ -1,6 +1,4 @@
 
-.linked "builtins.vml"
-.linked "fixed.vml"
 .data
 PI: .dword 3.141592653589793
 PI2: .dword 6.283185307179586
diff --git a/third_party/vml/Lib/shared/math64.vml b/third_party/vml/Lib/shared/math64.vml
index 30fb504c..f7169f5a 100644
--- a/third_party/vml/Lib/shared/math64.vml
+++ b/third_party/vml/Lib/shared/math64.vml
@@ -1,5 +1,4 @@
 
-.linked "fixed.vml"
 .data
 PI64_D: .dword 3.141592653589793
 PI2_D: .dword 6.283185307179586
diff --git a/third_party/vml/Lib/shared/matrix64.vml b/third_party/vml/Lib/shared/matrix64.vml
index 1fd33f60..f56825b5 100644
--- a/third_party/vml/Lib/shared/matrix64.vml
+++ b/third_party/vml/Lib/shared/matrix64.vml
@@ -1,5 +1,4 @@
 
-.linked "math64.vml"
 .data
 dbl_73430000: .dword 4607182418800017408
 dbl_73430001: .dword 0
diff --git a/third_party/vml/Lib/shared/parserexp.vml b/third_party/vml/Lib/shared/parserexp.vml
index 432eaa4b..9327b11d 100644
--- a/third_party/vml/Lib/shared/parserexp.vml
+++ b/third_party/vml/Lib/shared/parserexp.vml
@@ -1,5 +1,4 @@
 
-.linked "parserexpf.vml"
 .data
 g_pos: .word 0
 g_current: .word 0
diff --git a/third_party/vml/Lib/shared/pid.vml b/third_party/vml/Lib/shared/pid.vml
index 5a7ff2bd..b139ee9d 100644
--- a/third_party/vml/Lib/shared/pid.vml
+++ b/third_party/vml/Lib/shared/pid.vml
@@ -1,6 +1,4 @@
 
-.linked "ringbuf.vml"
-.linked "swtimer.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/printf.vml b/third_party/vml/Lib/shared/printf.vml
index e7d191cf..d382d47f 100644
--- a/third_party/vml/Lib/shared/printf.vml
+++ b/third_party/vml/Lib/shared/printf.vml
@@ -1,3 +1,4 @@
+
 .data
 L_35270013: .string "(null)"
 flt_35270100: .word 0
@@ -441,7 +442,7 @@ _L35270044:
         jz sta_35270041
         move R0 #0
         i2l R0 R0
-        move [R12-16] R0
+        movel [R12-16] R0
         move R0 #0
         move [R12-20] R0
         move R0 #0
@@ -525,13 +526,13 @@ _L35270060:
         move R0 #0
 _L35270061:
         move [R12-36] R0
-        move R0 [R12-16]
-        push R0
+        movel R0 [R12-16]
+        pushl R0
         move R0 #1
-        pop R1
+        popl R1
         shl R1 R0
-        move R0 R1
-        push R0
+        movel R0 R1
+        pushl R0
         move R0 [R12-28]
         push R0
         move R0 #32
@@ -551,16 +552,16 @@ _L35270064:
         move R0 [R12-36]
 _L35270065:
         i2l R0 R0
-        pop R1
+        popl R1
         or R1 R0
-        move R0 R1
-        move [R12-44] R0
-        move R0 [R12-44]
-        push R0
+        movel R0 R1
+        movel [R12-44] R0
+        movel R0 [R12-44]
+        pushl R0
         move R0 [R12+24]
         i2l R0 R0
-        pop R1
-        cmp R1 R0
+        popl R1
+        cmpl R1 R0
         jge _L35270068
         move R0 #0
         jmp _L35270069
@@ -569,27 +570,24 @@ _L35270068:
 _L35270069:
         cmp R0 #0
         jz sta_35270066
-        move R0 [R12-44]
-        push R0
+        movel R0 [R12-44]
+        pushl R0
         move R0 [R12+24]
         i2l R0 R0
-        pop R1
-        div R1 R0
-        move R0 R1
-        move [R12-52] R0
-        move R0 [R12-44]
-        push R0
-        move R0 [R12-52]
-        push R0
+        popl R1
+        divl R0 R1 R0
+        movel [R12-52] R0
+        movel R0 [R12-44]
+        pushl R0
+        movel R0 [R12-52]
+        pushl R0
         move R0 [R12+24]
         i2l R0 R0
-        pop R1
-        mul R1 R0
-        move R0 R1
-        pop R1
-        sub R1 R0
-        move R0 R1
-        move [R12-16] R0
+        popl R1
+        mull R0 R1 R0
+        popl R1
+        subl R0 R1 R0
+        movel [R12-16] R0
         move R0 [R12-28]
         push R0
         move R0 #32
@@ -605,12 +603,12 @@ _L35270073:
         jz sta_35270070
         move R0 [R12-20]
         push R0
-        move R0 [R12-52]
-        push R0
+        movel R0 [R12-52]
+        pushl R0
         move R0 #-1
-        pop R1
+        popl R1
         and R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         push R0
         move R0 [R12-28]
@@ -630,12 +628,12 @@ _L35270073:
 sta_35270070:
         move R0 [R12-24]
         push R0
-        move R0 [R12-52]
-        push R0
+        movel R0 [R12-52]
+        pushl R0
         move R0 #-1
-        pop R1
+        popl R1
         and R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         push R0
         move R0 [R12-28]
@@ -649,8 +647,8 @@ sta_35270070:
 sta_35270071:
         jmp sta_35270067
 sta_35270066:
-        move R0 [R12-44]
-        move [R12-16] R0
+        movel R0 [R12-44]
+        movel [R12-16] R0
 sta_35270067:
 sta_35270050:
         move R0 [R12-28]
@@ -660,11 +658,12 @@ sta_35270050:
         move R0 R1
         jmp sta_35270049
 sta_35270051:
-        move R0 [R12-16]
-        push R0
+        movel R0 [R12-16]
+        pushl R0
         move R0 #10
-        pop R1
-        cmp R1 R0
+        i2l R0 R0
+        popl R1
+        cmpl R1 R0
         jl _L35270074
         move R0 #0
         jmp _L35270075
@@ -675,7 +674,7 @@ _L35270075:
         jz _L35270076
         move R0 #48
         push R0
-        move R0 [R12-16]
+        movel R0 [R12-16]
         l2i R0 R0
         pop R1
         add R1 R0
@@ -691,7 +690,7 @@ _L35270078:
         move R0 #97
 _L35270079:
         push R0
-        move R0 [R12-16]
+        movel R0 [R12-16]
         l2i R0 R0
         pop R1
         add R1 R0
@@ -873,17 +872,17 @@ readLong:
         move [R12-12] R0
         move R0 [R12-12]
         i2l R0 R0
-        push R0
+        pushl R0
         move R0 #32
-        pop R1
+        popl R1
         shl R1 R0
-        move R0 R1
-        push R0
+        movel R0 R1
+        pushl R0
         move R0 [R12-8]
         i2l R0 R0
-        pop R1
+        popl R1
         or R1 R0
-        move R0 R1
+        movel R0 R1
         jmp L_35270091
 L_35270091:
         move R13 R12
@@ -926,19 +925,19 @@ readDouble:
         move [R12-12] R0
         move R0 [R12-12]
         i2l R0 R0
-        push R0
+        pushl R0
         move R0 #32
-        pop R1
+        popl R1
         shl R1 R0
-        move R0 R1
-        push R0
+        movel R0 R1
+        pushl R0
         move R0 [R12-8]
         i2l R0 R0
-        pop R1
+        popl R1
         or R1 R0
-        move R0 R1
-        move [R12-20] R0
-        move R0 [R12-20]
+        movel R0 R1
+        movel [R12-20] R0
+        movel R0 [R12-20]
         move R0 R12
         sub R0 #20
         move [R12-24] R0
@@ -964,11 +963,12 @@ sign64:
         push R12
         move R12 R13
         sub R13 #20
-        move R0 [R12+12]
-        push R0
+        movel R0 [R12+12]
+        pushl R0
         move R0 #0
-        pop R1
-        cmp R1 R0
+        i2l R0 R0
+        popl R1
+        cmpl R1 R0
         jl _L35270095
         move R0 #0
         jmp _L35270096
@@ -977,28 +977,28 @@ _L35270095:
 _L35270096:
         cmp R0 #0
         jz sta_35270093
-        move R0 [R12+12]
-        move R0 [R12+12]
-        neg R0 R0
-        move [R12-12] R0
+        movel R0 [R12+12]
+        movel R0 [R12+12]
+        negl R0 R0
+        movel [R12-12] R0
         move R0 [R12+20]
         move R8 R0
-        move R0 [R12-12]
-        push R0
+        movel R0 [R12-12]
+        pushl R0
         move R0 #-1
-        pop R1
+        popl R1
         and R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         move [R8] R0
         move R0 [R12+24]
         move R9 R0
-        move R0 [R12-12]
-        push R0
+        movel R0 [R12-12]
+        pushl R0
         move R0 #32
-        pop R1
+        popl R1
         shr R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         move [R9] R0
         move R0 #1
@@ -1007,22 +1007,22 @@ sta_35270093:
 sta_35270094:
         move R0 [R12+20]
         move R6 R0
-        move R0 [R12+12]
-        push R0
+        movel R0 [R12+12]
+        pushl R0
         move R0 #-1
-        pop R1
+        popl R1
         and R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         move [R6] R0
         move R0 [R12+24]
         move R7 R0
-        move R0 [R12+12]
-        push R0
+        movel R0 [R12+12]
+        pushl R0
         move R0 #32
-        pop R1
+        popl R1
         shr R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         move [R7] R0
         move R0 #0
@@ -2236,7 +2236,7 @@ sta_35270242:
         move [R12-44] R0
         move R0 #0
         i2l R0 R0
-        move [R12-52] R0
+        movel [R12-52] R0
         move R0 [R12-40]
         test R0 R0
         jz _L35270247
@@ -2273,7 +2273,7 @@ _L35270248:
         move R1 [R13+4]
         call readLong
         add R13 #8
-        move [R12-52] R0
+        movel [R12-52] R0
         move R0 [R12-16]
         push R0
         move R0 #2
@@ -2370,7 +2370,7 @@ _L35270259:
         move R0 R12
         sub R0 #108
         push R0
-        move R0 [R12-52]
+        movel R0 [R12-52]
         sub R13 #8
         movel @13 R0
         move R0 [R13+0]
@@ -2826,20 +2826,20 @@ _L35270328:
         move R0 [R12-40]
         cmp R0 #0
         jz sta_35270329
-        move R0 [R12-52]
-        push R0
+        movel R0 [R12-52]
+        pushl R0
         move R0 #-1
-        pop R1
+        popl R1
         and R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         move [R12-108] R0
-        move R0 [R12-52]
-        push R0
+        movel R0 [R12-52]
+        pushl R0
         move R0 #32
-        pop R1
+        popl R1
         shr R1 R0
-        move R0 R1
+        movel R0 R1
         l2i R0 R0
         move [R12-112] R0
         move R0 [R12-124]
@@ -2887,11 +2887,12 @@ sta_35270330:
         move R0 [R12-40]
         test R0 R0
         jz _L35270333
-        move R0 [R12-52]
-        push R0
+        movel R0 [R12-52]
+        pushl R0
         move R0 #0
-        pop R1
-        cmp R1 R0
+        i2l R0 R0
+        popl R1
+        cmpl R1 R0
         je _L35270335
         move R0 #0
         jmp _L35270336
diff --git a/third_party/vml/Lib/shared/printx.vml b/third_party/vml/Lib/shared/printx.vml
index 1d7844c6..68675de6 100644
--- a/third_party/vml/Lib/shared/printx.vml
+++ b/third_party/vml/Lib/shared/printx.vml
@@ -1,7 +1,4 @@
 
-.linked "builtins.vml"
-.linked "wchar.vml"
-.linked "uchar.vml"
 .data
 L_35090000: .string "0x"
 
diff --git a/third_party/vml/Lib/shared/ringbuf.vml b/third_party/vml/Lib/shared/ringbuf.vml
index 8809fd6c..3baca299 100644
--- a/third_party/vml/Lib/shared/ringbuf.vml
+++ b/third_party/vml/Lib/shared/ringbuf.vml
@@ -1,5 +1,4 @@
 
-.linked "swtimer.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/scanf.vml b/third_party/vml/Lib/shared/scanf.vml
index 70a51a52..8d04243a 100644
--- a/third_party/vml/Lib/shared/scanf.vml
+++ b/third_party/vml/Lib/shared/scanf.vml
@@ -1,5 +1,4 @@
 
-.linked "util.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/scheduler.vml b/third_party/vml/Lib/shared/scheduler.vml
index f4ce4bd2..17f57a6f 100644
--- a/third_party/vml/Lib/shared/scheduler.vml
+++ b/third_party/vml/Lib/shared/scheduler.vml
@@ -1,5 +1,4 @@
 
-.linked "swtimer.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/signal.vml b/third_party/vml/Lib/shared/signal.vml
index 7ed8b76f..310536fa 100644
--- a/third_party/vml/Lib/shared/signal.vml
+++ b/third_party/vml/Lib/shared/signal.vml
@@ -555,6 +555,173 @@ L_79280038:
         pop R12
         pop R15
         ret
+kalman_update:
+; -------------------------------------------
+; source   : int kalman_update(int* state,int measurement)
+; function : kalman_update
+; param   : int* state
+; param   : int measurement
+; return   : int
+; --------------------------------------------
+kalman_update:
+        push R15
+        push R12
+        move R12 R13
+        sub R13 #36
+        move R0 [R12+12]
+        push R0
+        move R0 #0
+        pop R1
+        cmp R1 R0
+        je _L79280041
+        move R0 #0
+        jmp _L79280042
+_L79280041:
+        move R0 #1
+_L79280042:
+        cmp R0 #0
+        jz sta_79280039
+        move R0 [R12+16]
+        jmp L_79280043
+sta_79280039:
+sta_79280040:
+        move R0 [R12+12]
+        push R0
+        move R0 #0
+        mul R0 #4
+        pop R8
+        add R0 R8
+        move R0 [R0]
+        move [R12-8] R0
+        move R0 [R12+12]
+        push R0
+        move R0 #1
+        mul R0 #4
+        pop R9
+        add R0 R9
+        move R0 [R0]
+        move [R12-12] R0
+        move R0 [R12+12]
+        push R0
+        move R0 #2
+        mul R0 #4
+        pop R11
+        add R0 R11
+        move R0 [R0]
+        move [R12-16] R0
+        move R0 [R12+12]
+        push R0
+        move R0 #3
+        mul R0 #4
+        pop R6
+        add R0 R6
+        move R0 [R0]
+        move [R12-20] R0
+        move R0 [R12-12]
+        push R0
+        move R0 [R12-16]
+        pop R1
+        add R1 R0
+        move R0 R1
+        move [R12-12] R0
+        move R0 [R12-12]
+        push R0
+        move R0 [R12-20]
+        pop R1
+        add R1 R0
+        move R0 R1
+        move [R12-24] R0
+        move R0 [R12-24]
+        push R0
+        move R0 #0
+        pop R1
+        cmp R1 R0
+        jg _L79280046
+        move R0 #0
+        jmp _L79280047
+_L79280046:
+        move R0 #1
+_L79280047:
+        cmp R0 #0
+        jz sta_79280044
+        move R0 [R12-12]
+        push R0
+        move R0 #1000
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 [R12-24]
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-28] R0
+        move R0 [R12-8]
+        push R0
+        move R0 [R12-28]
+        push R0
+        move R0 [R12+16]
+        push R0
+        move R0 [R12-8]
+        pop R1
+        sub R1 R0
+        move R0 R1
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #1000
+        pop R1
+        div R1 R0
+        move R0 R1
+        pop R1
+        add R1 R0
+        move R0 R1
+        move [R12-8] R0
+        move R0 #1000
+        push R0
+        move R0 [R12-28]
+        pop R1
+        sub R1 R0
+        move R0 R1
+        push R0
+        move R0 [R12-12]
+        pop R1
+        mul R1 R0
+        move R0 R1
+        push R0
+        move R0 #1000
+        pop R1
+        div R1 R0
+        move R0 R1
+        move [R12-12] R0
+sta_79280044:
+sta_79280045:
+        move R0 [R12-8]
+        move R7 R0
+        move R0 [R12+12]
+        push R0
+        move R0 #0
+        mul R0 #4
+        pop R5
+        add R0 R5
+        move [R0] R7
+        move R0 [R12-12]
+        move R4 R0
+        move R0 [R12+12]
+        push R0
+        move R0 #1
+        mul R0 #4
+        pop R3
+        add R0 R3
+        move [R0] R4
+        move R0 [R12-8]
+        jmp L_79280043
+L_79280043:
+        move R13 R12
+        pop R12
+        pop R15
+        ret
 lerp_table:
 ; -------------------------------------------
 ; source   : int lerp_table(int* x_table,int* y_table,int n,int x)
@@ -575,89 +742,89 @@ lerp_table:
         move R0 #0
         pop R1
         cmp R1 R0
-        je _L79280047
+        je _L79280056
         move R0 #0
-        jmp _L79280048
-_L79280047:
+        jmp _L79280057
+_L79280056:
         move R0 #1
-_L79280048:
+_L79280057:
         test R0 R0
-        jnz _L79280044
+        jnz _L79280053
         move R0 [R12+16]
         push R0
         move R0 #0
         pop R1
         cmp R1 R0
-        je _L79280049
+        je _L79280058
         move R0 #0
-        jmp _L79280050
-_L79280049:
+        jmp _L79280059
+_L79280058:
         move R0 #1
-_L79280050:
+_L79280059:
         test R0 R0
-        jz _L79280045
-_L79280044:
+        jz _L79280054
+_L79280053:
         move R0 #1
-        jmp _L79280046
-_L79280045:
+        jmp _L79280055
+_L79280054:
         move R0 #0
-_L79280046:
+_L79280055:
         test R0 R0
-        jnz _L79280041
+        jnz _L79280050
         move R0 [R12+20]
         push R0
         move R0 #2
         pop R1
         cmp R1 R0
-        jl _L79280051
+        jl _L79280060
         move R0 #0
-        jmp _L79280052
-_L79280051:
+        jmp _L79280061
+_L79280060:
         move R0 #1
-_L79280052:
+_L79280061:
         test R0 R0
-        jz _L79280042
-_L79280041:
+        jz _L79280051
+_L79280050:
         move R0 #1
-        jmp _L79280043
-_L79280042:
+        jmp _L79280052
+_L79280051:
         move R0 #0
-_L79280043:
+_L79280052:
         cmp R0 #0
-        jz sta_79280039
+        jz sta_79280048
         move R0 #0
-        jmp L_79280053
-sta_79280039:
-sta_79280040:
+        jmp L_79280062
+sta_79280048:
+sta_79280049:
         move R0 [R12+24]
         push R0
         move R0 [R12+12]
         push R0
         move R0 #0
         mul R0 #4
-        pop R8
-        add R0 R8
+        pop R2
+        add R0 R2
         move R0 [R0]
         pop R1
         cmp R1 R0
-        jle _L79280056
+        jle _L79280065
         move R0 #0
-        jmp _L79280057
-_L79280056:
+        jmp _L79280066
+_L79280065:
         move R0 #1
-_L79280057:
+_L79280066:
         cmp R0 #0
-        jz sta_79280054
+        jz sta_79280063
         move R0 [R12+16]
         push R0
         move R0 #0
         mul R0 #4
-        pop R9
-        add R0 R9
+        pop R1
+        add R0 R1
         move R0 [R0]
-        jmp L_79280053
-sta_79280054:
-sta_79280055:
+        jmp L_79280062
+sta_79280063:
+sta_79280064:
         move R0 [R12+24]
         push R0
         move R0 [R12+12]
@@ -669,19 +836,19 @@ sta_79280055:
         sub R1 R0
         move R0 R1
         mul R0 #4
-        pop R11
-        add R0 R11
+        pop R10
+        add R0 R10
         move R0 [R0]
         pop R1
         cmp R1 R0
-        jge _L79280060
+        jge _L79280069
         move R0 #0
-        jmp _L79280061
-_L79280060:
+        jmp _L79280070
+_L79280069:
         move R0 #1
-_L79280061:
+_L79280070:
         cmp R0 #0
-        jz sta_79280058
+        jz sta_79280067
         move R0 [R12+16]
         push R0
         move R0 [R12+20]
@@ -691,12 +858,12 @@ _L79280061:
         sub R1 R0
         move R0 R1
         mul R0 #4
-        pop R6
-        add R0 R6
+        pop R8
+        add R0 R8
         move R0 [R0]
-        jmp L_79280053
-sta_79280058:
-sta_79280059:
+        jmp L_79280062
+sta_79280067:
+sta_79280068:
         move R0 #0
         move [R12-8] R0
         move R0 [R12+20]
@@ -706,7 +873,7 @@ sta_79280059:
         sub R1 R0
         move R0 R1
         move [R12-12] R0
-sta_79280062:
+sta_79280071:
         move R2 #0
         move R0 [R12-8]
         push R0
@@ -718,14 +885,14 @@ sta_79280062:
         move R0 R1
         pop R1
         cmp R1 R0
-        jl _L79280064
+        jl _L79280073
         move R0 #0
-        jmp _L79280065
-_L79280064:
+        jmp _L79280074
+_L79280073:
         move R0 #1
-_L79280065:
+_L79280074:
         cmp R0 #0
-        jz sta_79280063
+        jz sta_79280072
         move R0 [R12-8]
         push R0
         move R0 [R12-12]
@@ -742,44 +909,44 @@ _L79280065:
         push R0
         move R0 [R12-16]
         mul R0 #4
-        pop R7
-        add R0 R7
+        pop R9
+        add R0 R9
         move R0 [R0]
         push R0
         move R0 [R12+24]
         pop R1
         cmp R1 R0
-        jle _L79280068
+        jle _L79280077
         move R0 #0
-        jmp _L79280069
-_L79280068:
+        jmp _L79280078
+_L79280077:
         move R0 #1
-_L79280069:
+_L79280078:
         cmp R0 #0
-        jz sta_79280066
+        jz sta_79280075
         move R0 [R12-16]
         move [R12-8] R0
-        jmp sta_79280067
-sta_79280066:
+        jmp sta_79280076
+sta_79280075:
         move R0 [R12-16]
         move [R12-12] R0
-sta_79280067:
-        jmp sta_79280062
-sta_79280063:
+sta_79280076:
+        jmp sta_79280071
+sta_79280072:
         move R0 [R12+12]
         push R0
         move R0 [R12-12]
         mul R0 #4
-        pop R5
-        add R0 R5
+        pop R11
+        add R0 R11
         move R0 [R0]
         push R0
         move R0 [R12+12]
         push R0
         move R0 [R12-8]
         mul R0 #4
-        pop R4
-        add R0 R4
+        pop R6
+        add R0 R6
         move R0 [R0]
         pop R1
         sub R1 R0
@@ -790,46 +957,46 @@ sta_79280063:
         move R0 #0
         pop R1
         cmp R1 R0
-        je _L79280072
+        je _L79280081
         move R0 #0
-        jmp _L79280073
-_L79280072:
+        jmp _L79280082
+_L79280081:
         move R0 #1
-_L79280073:
+_L79280082:
         cmp R0 #0
-        jz sta_79280070
+        jz sta_79280079
         move R0 [R12+16]
         push R0
         move R0 [R12-8]
         mul R0 #4
-        pop R3
-        add R0 R3
+        pop R5
+        add R0 R5
         move R0 [R0]
-        jmp L_79280053
-sta_79280070:
-sta_79280071:
+        jmp L_79280062
+sta_79280079:
+sta_79280080:
         move R0 [R12+16]
         push R0
         move R0 [R12-8]
         mul R0 #4
-        pop R2
-        add R0 R2
+        pop R7
+        add R0 R7
         move R0 [R0]
         push R0
         move R0 [R12+16]
         push R0
         move R0 [R12-12]
         mul R0 #4
-        pop R1
-        add R0 R1
+        pop R3
+        add R0 R3
         move R0 [R0]
         push R0
         move R0 [R12+16]
         push R0
         move R0 [R12-8]
         mul R0 #4
-        pop R10
-        add R0 R10
+        pop R4
+        add R0 R4
         move R0 [R0]
         pop R1
         sub R1 R0
@@ -841,8 +1008,8 @@ sta_79280071:
         push R0
         move R0 [R12-8]
         mul R0 #4
-        pop R8
-        add R0 R8
+        pop R2
+        add R0 R2
         move R0 [R0]
         pop R1
         sub R1 R0
@@ -858,8 +1025,8 @@ sta_79280071:
         pop R1
         add R1 R0
         move R0 R1
-        jmp L_79280053
-L_79280053:
+        jmp L_79280062
+L_79280062:
         move R13 R12
         pop R12
         pop R15
@@ -882,18 +1049,18 @@ deadband:
         move R0 [R12+16]
         pop R1
         cmp R1 R0
-        jg _L79280076
+        jg _L79280085
         move R0 #0
-        jmp _L79280077
-_L79280076:
+        jmp _L79280086
+_L79280085:
         move R0 #1
-_L79280077:
+_L79280086:
         cmp R0 #0
-        jz sta_79280074
+        jz sta_79280083
         move R0 [R12+12]
-        jmp L_79280078
-sta_79280074:
-sta_79280075:
+        jmp L_79280087
+sta_79280083:
+sta_79280084:
         move R0 [R12+12]
         push R0
         move R0 [R12+16]
@@ -901,21 +1068,21 @@ sta_79280075:
         neg R0 R0
         pop R1
         cmp R1 R0
-        jl _L79280081
+        jl _L79280090
         move R0 #0
-        jmp _L79280082
-_L79280081:
+        jmp _L79280091
+_L79280090:
         move R0 #1
-_L79280082:
+_L79280091:
         cmp R0 #0
-        jz sta_79280079
+        jz sta_79280088
         move R0 [R12+12]
-        jmp L_79280078
-sta_79280079:
-sta_79280080:
+        jmp L_79280087
+sta_79280088:
+sta_79280089:
         move R0 #0
-        jmp L_79280078
-L_79280078:
+        jmp L_79280087
+L_79280087:
         move R13 R12
         pop R12
         pop R15
@@ -941,76 +1108,76 @@ hysteresis:
         move R0 #0
         pop R1
         cmp R1 R0
-        je _L79280085
+        je _L79280094
         move R0 #0
-        jmp _L79280086
-_L79280085:
+        jmp _L79280095
+_L79280094:
         move R0 #1
-_L79280086:
+_L79280095:
         cmp R0 #0
-        jz sta_79280083
+        jz sta_79280092
         move R0 #0
-        jmp L_79280087
-sta_79280083:
-sta_79280084:
+        jmp L_79280096
+sta_79280092:
+sta_79280093:
         move R0 [R12+16]
         push R0
         move R0 [R12+20]
         pop R1
         cmp R1 R0
-        jg _L79280090
+        jg _L79280099
         move R0 #0
-        jmp _L79280091
-_L79280090:
+        jmp _L79280100
+_L79280099:
         move R0 #1
-_L79280091:
+_L79280100:
         cmp R0 #0
-        jz sta_79280088
+        jz sta_79280097
         move R0 #1
-        move R9 R0
+        move R1 R0
         move R0 [R12+12]
         push R0
         move R0 #0
         mul R0 #4
-        pop R11
-        add R0 R11
-        move [R0] R9
-        jmp sta_79280089
-sta_79280088:
+        pop R10
+        add R0 R10
+        move [R0] R1
+        jmp sta_79280098
+sta_79280097:
         move R0 [R12+16]
         push R0
         move R0 [R12+24]
         pop R1
         cmp R1 R0
-        jl _L79280094
+        jl _L79280103
         move R0 #0
-        jmp _L79280095
-_L79280094:
+        jmp _L79280104
+_L79280103:
         move R0 #1
-_L79280095:
+_L79280104:
         cmp R0 #0
-        jz sta_79280092
+        jz sta_79280101
         move R0 #0
-        move R6 R0
+        move R8 R0
         move R0 [R12+12]
         push R0
         move R0 #0
         mul R0 #4
-        pop R7
-        add R0 R7
-        move [R0] R6
-sta_79280092:
-sta_79280093:
-sta_79280089:
+        pop R9
+        add R0 R9
+        move [R0] R8
+sta_79280101:
+sta_79280102:
+sta_79280098:
         move R0 [R12+12]
         push R0
         move R0 #0
         mul R0 #4
-        pop R5
-        add R0 R5
+        pop R11
+        add R0 R11
         move R0 [R0]
-        jmp L_79280087
-L_79280087:
+        jmp L_79280096
+L_79280096:
         move R13 R12
         pop R12
         pop R15
diff --git a/third_party/vml/Lib/shared/softdouble.vml b/third_party/vml/Lib/shared/softdouble.vml
index d3bc1d18..1e57aae9 100644
--- a/third_party/vml/Lib/shared/softdouble.vml
+++ b/third_party/vml/Lib/shared/softdouble.vml
@@ -1,5 +1,4 @@
 
-.linked "math.vml"
 .data
 dbl_96590045: .dword 4746794007248502784
 dbl_96590046: .dword 4746794007248502784
diff --git a/third_party/vml/Lib/shared/statistics.vml b/third_party/vml/Lib/shared/statistics.vml
index cf64eb90..9b18f8d0 100644
--- a/third_party/vml/Lib/shared/statistics.vml
+++ b/third_party/vml/Lib/shared/statistics.vml
@@ -1,5 +1,4 @@
 
-.linked "fixed.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/statistics64.vml b/third_party/vml/Lib/shared/statistics64.vml
index adbf5e2f..4c2bdb64 100644
--- a/third_party/vml/Lib/shared/statistics64.vml
+++ b/third_party/vml/Lib/shared/statistics64.vml
@@ -1,5 +1,4 @@
 
-.linked "math64.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/sysinfo64.vml b/third_party/vml/Lib/shared/sysinfo64.vml
index 91edef90..2484f6cc 100644
--- a/third_party/vml/Lib/shared/sysinfo64.vml
+++ b/third_party/vml/Lib/shared/sysinfo64.vml
@@ -1,5 +1,4 @@
 
-.linked "sysinfo.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/uchar.vml b/third_party/vml/Lib/shared/uchar.vml
index 98551743..fd96d847 100644
--- a/third_party/vml/Lib/shared/uchar.vml
+++ b/third_party/vml/Lib/shared/uchar.vml
@@ -1,6 +1,4 @@
 
-.linked "uscanf.vml"
-.linked "ustring.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/uprintf.vml b/third_party/vml/Lib/shared/uprintf.vml
index 0bd638ce..79e03471 100644
--- a/third_party/vml/Lib/shared/uprintf.vml
+++ b/third_party/vml/Lib/shared/uprintf.vml
@@ -1,8 +1,4 @@
 
-.linked "uchar.vml"
-.linked "printf.vml"
-.linked "uscanf.vml"
-.linked "wprintf.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/uscanf.vml b/third_party/vml/Lib/shared/uscanf.vml
index bc775108..84ffc8c8 100644
--- a/third_party/vml/Lib/shared/uscanf.vml
+++ b/third_party/vml/Lib/shared/uscanf.vml
@@ -1,6 +1,4 @@
 
-.linked "uchar.vml"
-.linked "printf.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/ustring.vml b/third_party/vml/Lib/shared/ustring.vml
index 5afa8507..b2ff17f9 100644
--- a/third_party/vml/Lib/shared/ustring.vml
+++ b/third_party/vml/Lib/shared/ustring.vml
@@ -1,5 +1,4 @@
 
-.linked "uchar.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/wchar.vml b/third_party/vml/Lib/shared/wchar.vml
index 00deb313..48f9e828 100644
--- a/third_party/vml/Lib/shared/wchar.vml
+++ b/third_party/vml/Lib/shared/wchar.vml
@@ -1,6 +1,4 @@
 
-.linked "wscanf.vml"
-.linked "wstring.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/wprintf.vml b/third_party/vml/Lib/shared/wprintf.vml
index 7a893c0b..7bf0b8d8 100644
--- a/third_party/vml/Lib/shared/wprintf.vml
+++ b/third_party/vml/Lib/shared/wprintf.vml
@@ -1,6 +1,4 @@
 
-.linked "wchar.vml"
-.linked "printf.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/wscanf.vml b/third_party/vml/Lib/shared/wscanf.vml
index bb35ac66..14995b01 100644
--- a/third_party/vml/Lib/shared/wscanf.vml
+++ b/third_party/vml/Lib/shared/wscanf.vml
@@ -1,6 +1,4 @@
 
-.linked "wchar.vml"
-.linked "printf.vml"
 .data
 
 .text
diff --git a/third_party/vml/Lib/shared/wstring.vml b/third_party/vml/Lib/shared/wstring.vml
index 713cae3d..65c7743d 100644
--- a/third_party/vml/Lib/shared/wstring.vml
+++ b/third_party/vml/Lib/shared/wstring.vml
@@ -1,5 +1,4 @@
 
-.linked "wchar.vml"
 .data
 
 .text
